File: services/api/app/features/import_export/schemas.py
from datetime import date

from pydantic import BaseModel


class ImportSummary(BaseModel):
    inserted: int
    updated: int
    deleted: int


class ImportRowError(BaseModel):
    row: int
    messages: list[str]


class PendingDeletion(BaseModel):
    row_id: str
    description: str
    spent_on: date

File: frontend/features/import-export/schemas.ts
import { z } from "zod"

export const importFileSchema = z.object({
  file: z
    .instanceof(File, { error: "Choose a file to import" })
    .refine((file) => /\.(xlsx|xls)$/i.test(file.name), {
      message: "File must be .xlsx or .xls",
    }),
})

export type ImportFileFormValues = z.infer<typeof importFileSchema>

export type ImportSummary = {
  inserted: number
  updated: number
  deleted: number
}

export type ImportRowError = {
  row: number
  messages: string[]
}

export type PendingDeletion = {
  row_id: string
  description: string
  spent_on: string
}

export type ImportConfirmationRequired = {
  requires_confirmation: true
  deleted: PendingDeletion[]
}
